File: src/main/java/io/poshidaev/toda/entity/Goal.java
package io.poshidaev.toda.entity;

import lombok.*;

import javax.persistence.*;
import java.util.List;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Goal extends SingleIdEntity {

    public Goal(Long id, String text, GoalType type, Integer trialsCount, List<Topic> topics) {
        this.id = id;
        this.text = text;
        this.type = type;
        this.trialsCount = trialsCount;
        this.topics = topics;
    }

    @Column(nullable = false)
    private String text;

    @Enumerated(EnumType.ORDINAL)
    private GoalType type;

    private Integer trialsCount;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(name = "tasks_to_topics",
            joinColumns = {@JoinColumn(name = "goal_id", referencedColumnName = "id")},
            inverseJoinColumns = {@JoinColumn(name = "topic_id", referencedColumnName = "text", columnDefinition = "text")}
    )
    private List<Topic> topics;

    @OneToMany(fetch = FetchType.LAZY, mappedBy = "goal")
    private List<GoalApproach> approaches;
}

File: src/main/kotlin/io/poshidaev/toda/entity/Goal.kt
package io.poshidaev.toda.entity

import javax.persistence.*

@Entity
class Goal : SingleIdEntity{

    constructor(id: Long?, text: String?, type: GoalType, trialsCount: Int, topics: List<Topic>, approaches: List<GoalApproach>) : super(id) {
        this.text = text
        this.type = type
        this.trialsCount = trialsCount
        this.topics = topics
        this.approaches = approaches
    }


    constructor(id: Long?, text: String?) : super(id) {
        this.text = text
    }

    constructor(id: Long?, text: String?, type: GoalType, trialsCount: Int, topics: List<Topic>?) : super(id) {
        this.text = text
        this.type = type
        this.trialsCount = trialsCount
        this.topics = topics
    }

    @Column(nullable = false)
    var text: String? = null

    @Enumerated(EnumType.ORDINAL)
    var type: GoalType = GoalType.DAILY

    var trialsCount: Int = 1

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(name = "tasks_to_topics",
            joinColumns = [JoinColumn(name = "goal_id", referencedColumnName = "id")],
            inverseJoinColumns = [JoinColumn(name = "topic_id", referencedColumnName = "text", columnDefinition = "text")])
    var topics: List<Topic>? = mutableListOf()

    @OneToMany(fetch = FetchType.LAZY, mappedBy = "goal")
    var approaches: List<GoalApproach>? = mutableListOf()

}